### DISCUSSION TITLE
Question re: `window.titleBarStyle` logic

### DISCUSSION THREAD

**Discussion by colin-grant-work** [eclipse-theia/theia#10633] [Q&A] [answered]
@msujew, I wonder if you could help me understand the motivation for the communication between the IPC Renderer and IPC Main in electron when handling the `titleBarStyle` logic. I remember that it had something to do with issues with reading preferences, or maybe a multi-window context when all renderers share the same preference file, but I don't remember exactly the problem it solves or why it was the best solution.

In particular, I'm not sure I understand the mechanics of Electron's startup that cause the problem that Vince describes [here](https://github.com/eclipse-theia/theia/pull/10044#pullrequestreview-750327329). My naive understanding of how it _could_ work would be something like this

1. Start the application
2. Wait until preferences are ready to do anything related to menus
3. Read the preferences for `window.titleBarStyle` and `window.menuBarVisibility` and create the application menus appropriately, depending on platform.

I wouldn't _expect_ Electron itself to be doing anything between (1) and (3) that would interfere, but it sounds like it is in fact doing something?

**Comment by msujew** [comment, marked as answer]
@colin-grant-work Well, sadly Electron is doing something that interferes between 1 and 3. And it's basically all due to the fact that we have to decide on window startup whether we want it to have a frame (i.e. a native OS title and menu bar) or not.

At the point the preferences are ready, the window is already up and running and we cannot change the frame. Therefore we need to communicate to the electron backend, that the _next time_ we start the app, the window should be created with/without its frame. Hence, the electron app is in charge of the frame state of all windows and we query it when trying to identify whether we should create a browser-like menu title bar or to hide the top-panel. We cannot trust the actual preference, since it is independent of whether we currently are running frameless or not.

**Reply by colin-grant-work** [reply]
Thanks, I didn't realize that Electron needed to know which frame frame to use when starting up, but that does explain what's going on.

For reference, if anyone is curious, here's where ElectronMain decides which frame to use.

https://github.com/eclipse-theia/theia/blob/49524689f2e6bbada61fb5a429edebcc8e554d72/packages/core/src/electron-main/electron-main-application.ts#L372-L381
